feat: #2388: load balancer documentation when using the proxy via the operaor. (#2393)
kroxylicious/kroxylicious: docs/modules/configuring/proc-virtualkafkacluster-loadbalancer-dns-config.adoc

// file included in the following:
//
// con-virtualkafkafkacluster-overview.adoc

[id='proc-virtualkafkacluster-loadbalancer-dns-config-{context}']

= Configuring DNS (load balancer ingress type only)

[role="_abstract"]

When using the `loadBalancer` ingress type, it is the responsibility of the user to ensure that both the `bootstrapAddress` and the names generated from `advertisedBrokerAddressPattern` resolve to the *external address* of the Kubernetes Service underlying the load balancer on the *network where the off-cluster applications run*.

.Prerequisites

* Kroxylicious Operator is installed.
* `KafkaProxy`, `VirtualKafkaCluster`, and `KafkaProxyIngress` resources deployed.
* `VirtualKafkaCluster`, and `KafkaProxyIngress` configured for use of a `loadBalancer`.
* Ability to configure DNS on the network where the off-cluster applications run.

.Procedure

.  Run the following command to discover the *external address* being used by the load balancer.
+
[source,shell]
----

> kubectl get service -n <namespace> <proxy-name>-sni -o=jsonpath='{.status.loadBalancer.ingress[0].hostname}'

Reviewer: some implementations will have an `ip` field not a `hostname`, maybe we should peel it back to `.status.loadBalancer.ingress[0]` which will display either of them:

minikube:
```
kubectl get service -nmy-proxy simple-sni -o=jsonpath='{.status.loadBalancer.ingress[0]}'
{"ip":"127.0.0.1","ipMode":"VIP"}
```

(on that note, minikube needs a tunnel before it has an ip set on the Service, I wonder if these docs should cater to the development k8s environments? I'm imagining tire kickers using a local env first)

.Prerequisites

* Kroxylicious Operator is installed.
* `KafkaProxy`, `VirtualKafkaCluster`, and `KafkaProxyIngress` resources deployed.
* `VirtualKafkaCluster` and `KafkaProxyIngress` configured for use of a `loadBalancer`.
* Ability to configure DNS on the network where the off-cluster applications run.
* Network traffic is permitted to flow from the application network run to the external addresses provided by the Kubernetes cluster.

.Procedure

ifndef::OpenShiftOnly[]
. If using Minikube as your Kubernetes environment, enable the {Minikube}/docs/commands/tunnel/[Minikube Load Balancer tunnel] by running the following command.
  Use a separate console window to do this as the command needs to stay running for the tunnel to work.
+
[source,shell]
----
minikube tunnel
----
endif::OpenShiftOnly[]
. Run the following command to discover the *external address* being used by the load balancer.
+
[source,shell]
----
kubectl get service -n <namespace> <proxy-name>-sni -o=jsonpath='{.status.loadBalancer.ingress[0]}'
----
+

> Replace `<namespace>` with the name of the Kubernetes namespace where the resources are deployed and replace `<proxy-name>` with the name of the `KafkaProxy` resource.
> 
> . Depending on your Kubernetes environment, the command will return an IP address, or a hostname.
>   This is the external address of the load balancer.

Reviewer 2: ```suggestion
Replace `<namespace>` with the name of the Kubernetes namespace where the resources are deployed and replace `<proxy-name>` with the name of the `KafkaProxy` resource.
+
Depending on your Kubernetes environment, the command returns an IP address or a hostname.
This is the external address of the load balancer.
```

. Configure your DNS so that the bootstrap and broker names resolve to the external address.
+
Assuming a `bootstrapAddress` of `$(virtualClusterName).kafkaproxy.example.com`, an `advertisedBrokerAddressPattern` of `broker-$(nodeId).$(virtualClusterName).kafkaproxy.example.com`,

> a kafka cluster uses nodeIDs (0-2) and a virtual cluster name of `my-cluster`, the following DNS mappings would need to be made:

Reviewer 2: ```suggestion
a Kafka cluster uses nodeIDs (0-2), and a virtual cluster name of `my-cluster`, the following DNS mappings are listed:
```

+
[source,text]
----
my-cluster.kafkaproxy.example.com => <external address>
broker-0.my-cluster.kafkaproxy.example.com => <external address>
broker-1.my-cluster.kafkaproxy.example.com => <external address>
broker-2.my-cluster.kafkaproxy.example.com => <external address>
----
+

> The exact steps will vary depending on your environment. Talk to your Network Administrator.

Reviewer 2: ```suggestion
The exact steps vary by environment and network setup.
```

> . Confirm that the names resolve from the application network.

Reviewer 2: ```suggestion
. Confirm that the names resolve from the application network:
```
